2263 Get attached documents for recap.email

albertisfu opened the pull request:
Added support to get attached documents from NEFs. 

The process is the following:
- Detect if a notification contains attachments
- If there are attachments, get the attachment page and parse it
- Get attachments metadata
- Merge attachments into the docket entry
- These attachments are included within the docket alert and webhook.

Since we are reusing existing methods like `merge_attachment_page_data` the attachment page HTML is stored in S3. So that in case we need to debug something we can check the HTML file. We don't need to login into PACER to see the attachment page but seems that it's only available as long as documents are available. Once the documents are downloaded or after 15 days (expiration of magic links) the attachment page is no longer available.

Docket Alerts with attachments look like these:

<img width="1042" alt="Screen Shot 2022-08-20 at 2 07 16" src="https://user-images.githubusercontent.com/486004/185721314-eabc0b3c-7251-4274-b37d-f97fd559f117.png">

```
Your account on CourtListener just received its first email from PACER related to the case below.
You''ve been subscribed to this case because your account has auto-subscribe turned on.

**************************
CourtListener Docket Alert
**************************

1 New Entry in IN RE: NEW YORK CITY POLICING DURING SUMMER 2020 DEMONSTRATIONS (1:20-cv-08924)
District Court, S.D. New York
~~~
View Docket: https://www.courtlistener.com/docket/32/in-re-new-york-city-policing-during-summer-2020-demonstrations/?order_by=desc

Document Number: 721
Date Filed: Aug 8, 2022
LETTER addressed to Magistrate Judge Gabriel W. Gorenstein from Lillian Marquez
dated August 8, 2022 re: Request to Deem RFAs Admitted and Compel Discovery
Responses. Document filed by People of The State of New York. (Attachments: (1)
Exhibit A Objections and Responses, (2) Exhibit B Chart, (3) Exhibit C Suppl.
Objections and Responses, (4) Exhibit D Email).(Marquez, Lillian)
Download PDF from RECAP with PACER fallback: https://www.courtlistener.com/docket/32/721/in-re-new-york-city-policing-during-summer-2020-demonstrations/?redirect_to_download=True
Document Number: 721
Date Filed: Aug 8, 2022
Exhibit A Objections and Responses
Download PDF from RECAP with PACER fallback: https://www.courtlistener.com/docket/32/721/1/in-re-new-york-city-policing-during-summer-2020-demonstrations/?redirect_to_download=True
Document Number: 721
Date Filed: Aug 8, 2022
Exhibit B Chart
Download PDF from RECAP with PACER fallback: https://www.courtlistener.com/docket/32/721/2/in-re-new-york-city-policing-during-summer-2020-demonstrations/?redirect_to_download=True
Document Number: 721
Date Filed: Aug 8, 2022
Exhibit C Suppl. Objections and Responses
Download PDF from RECAP with PACER fallback: https://www.courtlistener.com/docket/32/721/3/in-re-new-york-city-policing-during-summer-2020-demonstrations/?redirect_to_download=True
Document Number: 721
Date Filed: Aug 8, 2022
Exhibit D Email
Download PDF from RECAP with PACER fallback: https://www.courtlistener.com/docket/32/721/4/in-re-new-york-city-policing-during-summer-2020-demonstrations/?redirect_to_download=True
```

Attachments are also sent to webhooks.

Some changes and more examples were added on [Juriscraper](https://github.com/freelawproject/juriscraper/pull/546).

Let me know what you think.

Thanks

mlissner commented on cl/recap/mergers.py:
```diff
@@ -1399,3 +1405,56 @@ def process_orphan_documents(
             pass
+
+
+def add_att_report_for_recap_email(
```
This is copying a lot of code from elsewhere. Do you think we could generalize the other functions instead of duplicating them?

albertisfu replied:
Yeah, I've removed this method.

Now after getting the attachment page, we create a PQ using `make_attachment_pq_object` to process the attachment page, I tweaked that method to also accept the HTML attachment page (in case we get it from the notification URL, without querying PACER).

And then we use `process_recap_attachment` to merge the attached documents, I tweaked the method because here we need to return the attachments that were added so we can try to download them.

I tried to call these tasks asynchronously using a chain but I found some problems because we need to send the alerts and webhooks after we'd finished getting all the documents, and using the chain wasn't working well.  I think this might be possible using a combination of celery chains, groups, and chords, but it might complicate things a bit.

mlissner commented on cl/recap/tests.py:
```diff
@@ -1855,11 +1855,18 @@ def test_create_from_idb_chunk(self) -> None:
+    if read_file:
+        with open(test_dir / message_id, encoding="utf-8") as file:
+            text = file.read()
+            file.close()
+            return text
```
You're using a context handler — it'll close the file itself, no?

albertisfu replied:
yeah, removed!

mlissner reviewed (commented):
No major surprises here. I guess I'd like to see if we can remove `add_att_report_for_recap_email` by combining it with other code.

I'll let you update for the latest juriscraper too and then I'll have more comments.

mlissner commented:
A couple other comments I forgot to make before. We've never sent attachment info as parts of alerts before, so we need to tweak the email to make it make more sense. 

Two things I think would help:

 - [ ] For both versions of the email, we need to start displaying the attachment number. The easiest way to do this is with a simple hyphen, so some code that says, roughly:

    ```
    {{ doc.number }}{% if doc.is_attachment %}-{{ doc.att_num }}{% endif %}
    ```

    That'd probably be fine, and it's easy. But in the screenshot you have above, where it just shows doc 721 over and over — that's no good. It's common to show attachments as 721-1, 721-2, etc.

 - In the plaintext email, we need a little whitespace between the entries. They're *really* crammed. Something is needed there to make those emails more legible.

albertisfu commented:
Thanks for your comments and suggestions, I've applied the changes.

I added a fallback just in case the attachment page from the notifications link (not pacer login required) is not available for some reason, now we can also try to get the attachment page after login into PACER since it's free to consult the attachment page. In most cases, it's supposed we are going to save the query into PACER to consult the attachment page since the anonymous attachment page will be available.

 I applied the hyphen to show the attachment number for every doc in the notification:

<img width="1175" alt="Screen Shot 2022-08-25 at 17 22 54" src="https://user-images.githubusercontent.com/486004/186718267-ce00a07f-fc9f-46f5-b84e-00c56ddcbae4.png">

For the text plain version, I also added a space between documents:

```
Your account on CourtListener just received its first email from PACER related to the case below.
You''ve been subscribed to this case because your account has auto-subscribe turned on.

**************************
CourtListener Docket Alert
**************************

1 New Entry in Grant v. Gusman (2:17-cv-02797)
District Court, E.D. Louisiana.
~~~
View Docket: https://www.courtlistener.com/docket/75/grant-v-gusman/?order_by=desc

Document Number: 251
Date Filed: Aug 11, 2022
SCHEDULING ORDER: Final Pretrial Conference set for 2/23/2023 04:00 PM before
Chief Judge Nannette Jolivette Brown. Jury Trial set for 3/20/2023 09:00 AM
before Chief Judge Nannette Jolivette Brown. Issued by Clerk. (Attachments: (1)
Pretrial Notice Form, (2) Standing Order)(mp)
Download PDF from RECAP with PACER fallback: https://www.courtlistener.com/docket/75/251/grant-v-gusman/?redirect_to_download=True

Document Number: 251-1
Date Filed: Aug 11, 2022
Pretrial Notice Form
Download PDF from RECAP: https://dev-com-courtlistener-storage.s3.amazonaws.com/recap/dev.gov.uscourts.laed.195449/gov.uscourts.laed.195449.251.1.pdf

Document Number: 251-2
Date Filed: Aug 11, 2022
Standing Order
Download PDF from RECAP: https://dev-com-courtlistener-storage.s3.amazonaws.com/recap/dev.gov.uscourts.laed.195449/gov.uscourts.laed.195449.251.2.pdf




This alert was sent because you subscribed to this docket with your account on CourtListener.com.
To disable this alert unsubscribe here: https://www.courtlistener.com/alert/docket/unsubscribe/wHr6fB1Gpsjh1D8ih0sHOnU5st1Q7KuJXEF0Gqfy/

************************
This alert brought to you by the 501(c)(3) non-profit Free Law Project

 - Blog: https://free.law
 - Twitter: http://twitter.com/freelawproject
 - Donate: https://www.courtlistener.com/donate/?referrer=pt-email

Please donate to support our work.
```
Let me know what you think.

mlissner commented on cl/recap/tasks.py:
```diff
@@ -564,13 +566,16 @@ def process_recap_attachment(
     :param pk: The primary key of the processing queue item you want to work on
     :param tag_names: A list of tag names to add to all items created or
     modified in this function.
+    :param return_rds_affected: Whether to return a list of attachments
+    documents that were added or not.
```
I can see you're trying to be surgical here and make the smallest change possible, but it's better to make a big change than to pile on weird code that changes the return type based on the args passed. 

With that in mind, what'd suggest is simply updating the return so it always returns everything:

```
-> Optional[Tuple[int, str, list[RECAPDocument]]]:
```

That'll require updating anywhere that uses the return types from this function, but that shouldn't be too hard, right?

albertisfu replied:
Changed, great thanks. This is good advice! sometimes I'm not sure if it's better to do a major refactor or try to not modify much of the existing method, now I know that it's better to make big changes to avoid fragmented code.

The good thing is that they were not necessary further changes since process_recap_attachment returns are not being used in any other place than the new one.

mlissner replied:
Unless we're doing a hot fix or it's really urgent, always write the best code you can. I try to never compromise on code quality. We're in this for the long haul. 

(Some bugs are just too big to prioritize though, I'll admit that.)

mlissner commented on cl/alerts/templates/docket_alert_email.html:
```diff
@@ -73,7 +73,7 @@ <h3 class="alt bottom" style="font-size: 1.5em; font-weight: normal; line-height
         {% for rd in de.recap_documents.all %}
           <tr>
-            <td style="text-align: center">{{ de.entry_number }}</td>
+            <td style="text-align: center">{{ de.entry_number }}{% if rd.attachment_number  %}-{{ rd.attachment_number }}{% endif %}</td>
```
OK, I'll admit I'm a bit proud how close my pseudo-code was!

albertisfu replied:
yeah, very accurate!

mlissner commented on cl/corpus_importer/tasks.py:
```diff
@@ -1508,6 +1508,7 @@ def make_attachment_pq_object(
     attachment_report: AttachmentPage,
     rd_pk: int,
     user_pk: int,
+    att_report_text: str = None,
```
Shouldn't this be either `str | None` or `Optional[str]`?

albertisfu replied:
Yeah, I changed to str | None = None

Thanks!

mlissner reviewed (changes requested):
Couple little things. Home stretch.

albertisfu commented:
Thanks! These changes are done.

Since we updated juriscraper here, is it not needed to do a bump version here to build the new image, right?  that's now done automatically?

mlissner commented on cl/recap/tasks.py:
```diff
@@ -564,13 +566,15 @@ def process_recap_attachment(
     :param pk: The primary key of the processing queue item you want to work on
     :param tag_names: A list of tag names to add to all items created or
     modified in this function.
+    :param return_rds_affected: Whether to return a list of attachments
+    documents that were added or not.
```
This needs to go.

albertisfu replied:
Done.

mlissner commented:
One last tweak on the params, then please merge.

Juriscraper 2.5.13 is released on PyPi: https://pypi.org/project/juriscraper/. Does that answer your question?

albertisfu commented:
Great, thanks.  

Well, I remember that before auto-deploy CI, we had to update the version on the version.txt, now is not needed, right?

mlissner commented:
version.txt is no more. Auto-merge is set to go!
